### PR TITLE
Use foldFree to implement the free monad interpreters

### PR DESCRIPTION
**Pull request by ruuda**
As suggested earlier by @rightfold, we can simplify the matching with `foldFree`. This is indeed a nice cleanup!

In the tests, when using `foldFree` we need a monad instance, but fortunately `Monoid a => Monad ((,) a)` provides one, where we can instantiate `a = [ActionFlat]`. That allows removing even more boilerplate! The downside is that in order to use the instance, we need to swap the elements in the returned tuple. Not a big deal, but it clutters the diff.

This pull request can be reviewed commit by commit. The first two commits are the important ones, the last two are cosmetic.

### CODE REVIEW

File: tests/Spec.hs [hunk 1]
```diff
@@ -67,51 +68,54 @@ data ActionFlat
...
   in
-    case action of
-      Pure result                          -> (result, [])
-      Free (TryIntegrate msg candi h)      -> prepend (h integrateResult) $ ATryIntegrate msg candi
-      Free (TryPromote prBranch headSha h) -> prepend (h pushResult) $ ATryPromote prBranch headSha
-      Free (LeaveComment pr body x)        -> prepend x $ ALeaveComment pr body
-      Free (IsReviewer username h)         -> prepend (h $ isReviewer username) $ AIsReviewer username
+    -- This relies on the monad instance `Monoid a => Monad ((,) a)`, which
+    -- relies on the monoid instance of list. At every step, we get to provide
+    -- a list, and the results get concatenated.
```
Comment (rightfold) [inline tests/Spec.hs]: If you find the monad instance (,) too obscure, you could use the isomorphic Writer data type from the transformers package.

```haskell
Monoid a => Monad (Writer a)
TryIntegrate msg candi h -> tell [ATryIntegrate msg candi] $> h integrateResult
```

What’s nice is that runWriter returns (a, [ActionFlat]) just like your original code.

Comment (ruuda) [reply tests/Spec.hs]: Great suggestion, implemented.

**Comment by ruuda** [comment]
@OpsBotPrime merge

**Comment by OpsBotPrime** [comment]
Rebased as a4856e288c7b67afdbf90044a23dfc7505ce8fe8, waiting for CI …

**Pull request merged**